BUG: `scipy/optimize/tests/test_hessian_update_strategy.py::TestHessianUpdateStrategy::test_initialize_catch_illegal` fails on PyPy3

### Describe your issue.

The following tests fails on PyPy3.10 7.3.16, apparently over a slightly different exception message:

```
FAILED scipy/optimize/tests/test_hessian_update_strategy.py::TestHessianUpdateStrategy::test_initialize_catch_illegal - AssertionError: Regex pattern did not match.
 Regex: "float\\(\\)\\ argument\\ must\\ be\\ a\\ string\\ or\\ a\\ real\\ number,\\ not\\ 'complex'"
 Input: "float() argument must be a string or a number, not 'complex'"
```

Note that "real" is missing.

Indeed:

```pytb
$ python3.10 -c 'float(complex(1,1))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: float() argument must be a string or a real number, not 'complex'
$ pypy3.10 -c 'float(complex(1,1))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: float() argument must be a string or a number, not 'complex'
```

### Reproducing Code Example

```python
python -m pytest scipy/optimize/tests/test_hessian_update_strategy.py::TestHessianUpdateStrategy::test_initialize_catch_illegal  # ;-)
```


### Error message

```shell
============================================================== FAILURES ===============================================================
_______________________________________ TestHessianUpdateStrategy.test_initialize_catch_illegal _______________________________________
[gw2] linux -- Python 3.10.14 /var/tmp/portage/dev-python/scipy-1.14.0/work/scipy-1.14.0-pypy3/install/usr/bin/pypy3

self = <scipy.optimize.tests.test_hessian_update_strategy.TestHessianUpdateStrategy testMethod=test_initialize_catch_illegal>

    def test_initialize_catch_illegal(self):
        ndims = 3
        # no complex allowed
        inits_msg_errtype = ((complex(3.14),
                              re.escape("float() argument must be a "
                                        "string or a real number, "
                                        "not 'complex'"),
                              TypeError),
    
                             (np.array([3.2, 2.3, 1.2]).astype(np.complex128),
                              "init_scale contains complex elements, "
                              "must be real.",
                              TypeError),
    
                             (np.array([[43, 24, 33],
                                        [24, 36, 44, ],
                                        [33, 44, 37, ]]).astype(np.complex128),
                              "init_scale contains complex elements, "
                              "must be real.",
                              TypeError),
    
                             # not square
                             (np.array([[43, 55, 66]]),
                              re.escape(
                                  "If init_scale is an array, it must have the "
                                  "dimensions of the hess/inv_hess: (3, 3)."
                                  " Got (1, 3)."),
                              ValueError),
    
                             # not symmetric
                             (np.array([[43, 24, 33],
                                        [24.1, 36, 44, ],
                                        [33, 44, 37, ]]),
                              re.escape("If init_scale is an array, it must be"
                                        " symmetric (passing scipy.linalg.issymmetric)"
                                        " to be an approximation of a hess/inv_hess."),
                              ValueError),
                             )
        for approx_type in ['hess', 'inv_hess']:
            for init_scale, message, errortype in inits_msg_errtype:
                # large min_{denominator,curvatur} makes it skip an update,
                # so we can retrieve our initial matrix
                quasi_newton = (BFGS(init_scale=init_scale),
                                SR1(init_scale=init_scale))
    
                for qn in quasi_newton:
                    qn.initialize(ndims, approx_type)
                    with pytest.raises(errortype, match=message):
>                       qn.update(np.ones(ndims), np.arange(ndims))

approx_type = 'hess'
errortype  = <class 'TypeError'>
init_scale = (3.14+0j)
inits_msg_errtype = (((3.14+0j),
  'float\\(\\)\\ argument\\ must\\ be\\ a\\ string\\ or\\ a\\ real\\ number,\\ '
  "not\\ 'complex'",
  <class 'TypeError'>),
 (array([3.2+0.j, 2.3+0.j, 1.2+0.j]),
  'init_scale contains complex elements, must be real.',
  <class 'TypeError'>),
 (array([[43.+0.j, 24.+0.j, 33.+0.j],
       [24.+0.j, 36.+0.j, 44.+0.j],
       [33.+0.j, 44.+0.j, 37.+0.j]]),
  'init_scale contains complex elements, must be real.',
  <class 'TypeError'>),
 (array([[43, 55, 66]]),
  'If\\ init_scale\\ is\\ an\\ array,\\ it\\ must\\ have\\ the\\ dimensions\\ '
  'of\\ the\\ hess/inv_hess:\\ \\(3,\\ 3\\)\\.\\ Got\\ \\(1,\\ 3\\)\\.',
  <class 'ValueError'>),
 (array([[43. , 24. , 33. ],
       [24.1, 36. , 44. ],
       [33. , 44. , 37. ]]),
  'If\\ init_scale\\ is\\ an\\ array,\\ it\\ must\\ be\\ symmetric\\ '
  '\\(passing\\ scipy\\.linalg\\.issymmetric\\)\\ to\\ be\\ an\\ '
  'approximation\\ of\\ a\\ hess/inv_hess\\.',
  <class 'ValueError'>))
message    = ('float\\(\\)\\ argument\\ must\\ be\\ a\\ string\\ or\\ a\\ real\\ number,\\ '
 "not\\ 'complex'")
ndims      = 3
qn         = <scipy.optimize._hessian_update_strategy.BFGS object at 0x0000564bd7769088>
quasi_newton = (<scipy.optimize._hessian_update_strategy.BFGS object at 0x0000564bd7769088>,
 <scipy.optimize._hessian_update_strategy.SR1 object at 0x0000564bd7769168>)
self       = <scipy.optimize.tests.test_hessian_update_strategy.TestHessianUpdateStrategy testMethod=test_initialize_catch_illegal>

scipy/optimize/tests/test_hessian_update_strategy.py:147: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <scipy.optimize._hessian_update_strategy.BFGS object at 0x0000564bd7769088>, delta_x = array([1., 1., 1.])
delta_grad = array([0, 1, 2])

    def update(self, delta_x, delta_grad):
        """Update internal matrix.
    
        Update Hessian matrix or its inverse (depending on how 'approx_type'
        is defined) using information about the last evaluated points.
    
        Parameters
        ----------
        delta_x : ndarray
            The difference between two points the gradient
            function have been evaluated at: ``delta_x = x2 - x1``.
        delta_grad : ndarray
            The difference between the gradients:
            ``delta_grad = grad(x2) - grad(x1)``.
        """
        if np.all(delta_x == 0.0):
            return
        if np.all(delta_grad == 0.0):
            warn('delta_grad == 0.0. Check if the approximated '
                 'function is linear. If the function is linear '
                 'better results can be obtained by defining the '
                 'Hessian as zero instead of using quasi-Newton '
                 'approximations.',
                 UserWarning, stacklevel=2)
            return
        if self.first_iteration:
            # Get user specific scale
            if isinstance(self.init_scale, str) and self.init_scale == "auto":
                scale = self._auto_scale(delta_x, delta_grad)
            else:
                scale = self.init_scale
    
            # Check for complex: numpy will silently cast a complex array to
            # a real one but not so for scalar as it raises a TypeError.
            # Checking here brings a consistent behavior.
            replace = False
            if np.size(scale) == 1:
                # to account for the legacy behavior having the exact same cast
>               scale = float(scale)
E               TypeError: float() argument must be a string or a number, not 'complex'

delta_grad = array([0, 1, 2])
delta_x    = array([1., 1., 1.])
replace    = False
scale      = (3.14+0j)
self       = <scipy.optimize._hessian_update_strategy.BFGS object at 0x0000564bd7769088>

scipy/optimize/_hessian_update_strategy.py:202: TypeError

During handling of the above exception, another exception occurred:

self = <scipy.optimize.tests.test_hessian_update_strategy.TestHessianUpdateStrategy testMethod=test_initialize_catch_illegal>

    def test_initialize_catch_illegal(self):
        ndims = 3
        # no complex allowed
        inits_msg_errtype = ((complex(3.14),
                              re.escape("float() argument must be a "
                                        "string or a real number, "
                                        "not 'complex'"),
                              TypeError),
    
                             (np.array([3.2, 2.3, 1.2]).astype(np.complex128),
                              "init_scale contains complex elements, "
                              "must be real.",
                              TypeError),
    
                             (np.array([[43, 24, 33],
                                        [24, 36, 44, ],
                                        [33, 44, 37, ]]).astype(np.complex128),
                              "init_scale contains complex elements, "
                              "must be real.",
                              TypeError),
    
                             # not square
                             (np.array([[43, 55, 66]]),
                              re.escape(
                                  "If init_scale is an array, it must have the "
                                  "dimensions of the hess/inv_hess: (3, 3)."
                                  " Got (1, 3)."),
                              ValueError),
    
                             # not symmetric
                             (np.array([[43, 24, 33],
                                        [24.1, 36, 44, ],
                                        [33, 44, 37, ]]),
                              re.escape("If init_scale is an array, it must be"
                                        " symmetric (passing scipy.linalg.issymmetric)"
                                        " to be an approximation of a hess/inv_hess."),
                              ValueError),
                             )
        for approx_type in ['hess', 'inv_hess']:
            for init_scale, message, errortype in inits_msg_errtype:
                # large min_{denominator,curvatur} makes it skip an update,
                # so we can retrieve our initial matrix
                quasi_newton = (BFGS(init_scale=init_scale),
                                SR1(init_scale=init_scale))
    
                for qn in quasi_newton:
                    qn.initialize(ndims, approx_type)
>                   with pytest.raises(errortype, match=message):
E                   AssertionError: Regex pattern did not match.
E                    Regex: "float\\(\\)\\ argument\\ must\\ be\\ a\\ string\\ or\\ a\\ real\\ number,\\ not\\ 'complex'"
E                    Input: "float() argument must be a string or a number, not 'complex'"

approx_type = 'hess'
errortype  = <class 'TypeError'>
init_scale = (3.14+0j)
inits_msg_errtype = (((3.14+0j),
  'float\\(\\)\\ argument\\ must\\ be\\ a\\ string\\ or\\ a\\ real\\ number,\\ '
  "not\\ 'complex'",
  <class 'TypeError'>),
 (array([3.2+0.j, 2.3+0.j, 1.2+0.j]),
  'init_scale contains complex elements, must be real.',
  <class 'TypeError'>),
 (array([[43.+0.j, 24.+0.j, 33.+0.j],
       [24.+0.j, 36.+0.j, 44.+0.j],
       [33.+0.j, 44.+0.j, 37.+0.j]]),
  'init_scale contains complex elements, must be real.',
  <class 'TypeError'>),
 (array([[43, 55, 66]]),
  'If\\ init_scale\\ is\\ an\\ array,\\ it\\ must\\ have\\ the\\ dimensions\\ '
  'of\\ the\\ hess/inv_hess:\\ \\(3,\\ 3\\)\\.\\ Got\\ \\(1,\\ 3\\)\\.',
  <class 'ValueError'>),
 (array([[43. , 24. , 33. ],
       [24.1, 36. , 44. ],
       [33. , 44. , 37. ]]),
  'If\\ init_scale\\ is\\ an\\ array,\\ it\\ must\\ be\\ symmetric\\ '
  '\\(passing\\ scipy\\.linalg\\.issymmetric\\)\\ to\\ be\\ an\\ '
  'approximation\\ of\\ a\\ hess/inv_hess\\.',
  <class 'ValueError'>))
message    = ('float\\(\\)\\ argument\\ must\\ be\\ a\\ string\\ or\\ a\\ real\\ number,\\ '
 "not\\ 'complex'")
ndims      = 3
qn         = <scipy.optimize._hessian_update_strategy.BFGS object at 0x0000564bd7769088>
quasi_newton = (<scipy.optimize._hessian_update_strategy.BFGS object at 0x0000564bd7769088>,
 <scipy.optimize._hessian_update_strategy.SR1 object at 0x0000564bd7769168>)
self       = <scipy.optimize.tests.test_hessian_update_strategy.TestHessianUpdateStrategy testMethod=test_initialize_catch_illegal>

scipy/optimize/tests/test_hessian_update_strategy.py:146: AssertionError
```


### SciPy/NumPy/Python version and system information

```shell
1.14.0 2.0.0 sys.version_info(major=3, minor=10, micro=14, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/include
    lib directory: /usr/lib64
    name: blas
    openblas configuration: unknown
    pc file directory: /usr/lib64/pkgconfig
    version: 3.12.0
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/include
    lib directory: /usr/lib64
    name: lapack
    openblas configuration: unknown
    pc file directory: /usr/lib64/pkgconfig
    version: 3.12.0
  pybind11:
    detection method: pkgconfig
    include directory: /include
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    args: -march=znver2, --param=l1-cache-size=32, --param=l1-cache-line-size=64,
      -O2, -pipe, -frecord-gcc-switches, -UNDEBUG
    commands: x86_64-pc-linux-gnu-gcc
    linker: ld.bfd
    linker args: -march=znver2, --param=l1-cache-size=32, --param=l1-cache-line-size=64,
      -O2, -pipe, -frecord-gcc-switches, -Wl,-O1, -Wl,--as-needed, -Wl,--defsym=__gentoo_check_ldflags__=0
    name: gcc
    version: 14.1.1
  c++:
    args: -march=znver2, --param=l1-cache-size=32, --param=l1-cache-line-size=64,
      -O2, -pipe, -frecord-gcc-switches, -UNDEBUG
    commands: x86_64-pc-linux-gnu-g++
    linker: ld.bfd
    linker args: -march=znver2, --param=l1-cache-size=32, --param=l1-cache-line-size=64,
      -O2, -pipe, -frecord-gcc-switches, -Wl,-O1, -Wl,--as-needed, -Wl,--defsym=__gentoo_check_ldflags__=0
    name: gcc
    version: 14.1.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    args: -march=znver2, --param=l1-cache-size=32, --param=l1-cache-line-size=64,
      -O2, -pipe, -frecord-gcc-switches
    commands: x86_64-pc-linux-gnu-gfortran
    linker: ld.bfd
    linker args: -march=znver2, --param=l1-cache-size=32, --param=l1-cache-line-size=64,
      -O2, -pipe, -frecord-gcc-switches, -Wl,-O1, -Wl,--as-needed, -Wl,--defsym=__gentoo_check_ldflags__=0
    name: gcc
    version: 14.1.1
  pythran:
    include directory: ../../../../../../../../usr/lib/pypy3.10/site-packages/pythran
    version: 0.16.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /usr/bin/pypy3
  version: '3.10'
```

--- a/scipy/optimize/tests/test_hessian_update_strategy.py
+++ b/scipy/optimize/tests/test_hessian_update_strategy.py
@@ -100,9 +100,8 @@ class TestHessianUpdateStrategy(TestCase):
         ndims = 3
         # no complex allowed
         inits_msg_errtype = ((complex(3.14),
-                              re.escape("float() argument must be a "
-                                        "string or a real number, "
-                                        "not 'complex'"),
+                              r"float\(\) argument must be a string or a "
+                              r"(real )?number, not 'complex'",
                               TypeError),
 
                              (np.array([3.2, 2.3, 1.2]).astype(np.complex128),

